charmbits/httpservice/server.go

```diff
@@ -0,0 +1,439 @@
+	httpOK := state.HTTPPort != 0
+	httpsOK := state.HTTPSPort != 0 && state.CertPEM != ""
+	h, err := srv.handlerInfo.handler(state.ArgData, srv.relationState)
+	if err != nil {
+		fb.addError(errgo.Notef(err, "cannot get handler"))
+		return
+	}
+	if srv.httpListener == nil && httpOK {
+		log.Printf("starting HTTP server on port %v", state.HTTPPort)
+		if err := srv.serveHTTP(state.HTTPPort, h); err != nil {
+			fb.addError(errgo.Notef(err, "cannot start HTTP server"))
+		}
+	}
+	if srv.httpsListener == nil && httpsOK {
+		log.Printf("starting HTTPS server on port %v", state.HTTPSPort)
+		if err := srv.serveHTTPS(state.HTTPSPort, state.CertPEM, h); err != nil {
+			fb.addError(errgo.Notef(err, "cannot start HTTPS server"))
+		}
+	}
+	srv.handler = h
+	srv.state = state
+	return
+}
+
+type handlerListener struct {
+	tomb tomb.Tomb
+	lis  net.Listener
+}
+
+func newHandlerListener(server *http.Server, lis net.Listener) *handlerListener {
+	hl := &handlerListener{
+		lis: lis,
+	}
+	hl.tomb.Go(func() error {
+		if err := server.Serve(lis); err != nil {
+			return errgo.Notef(err, "listener on %s died", lis.Addr())
+		}
+		return nil
+	})
+	return hl
+}
+
+func (hl *handlerListener) Kill() {
+	hl.lis.Close()
+}
+
+func (hl *handlerListener) Wait() error {
+	hl.tomb.Wait()
+	return nil
+}
+
+// needsRestart reports whether the server
+// needs a restart when the state is changed to the given
+// state.
+func (srv *server) needsRestart(state ServerState) bool {
+	// Note that when the mongodb addresses change, that
+	// doesn't necessarily imply we'll need a restart,
+	// as mongodb informs all clients of the new addresses
+	// as a matter of course. We'll store the addresses however,
+	// and they'll be used when reconnecting.
+	return state.HTTPPort == 0 ||
+		state.HTTPPort != srv.state.HTTPPort ||
+		state.HTTPSPort != srv.state.HTTPSPort ||
+		state.CertPEM != srv.state.CertPEM ||
+		!bytes.Equal(state.ArgData, srv.state.ArgData)
+}
+
+// setRelations sets the relation fields from the information in the
+// given state. If it returns ErrRestartNeeded, the handler
+// must be restarted.
+func (srv *server) setRelations(state ServerState) error {
+	h := srv.handlerInfo
+	if h.relationType == nil {
+		return nil
+	}
+	if !srv.relationState.IsValid() {
+		srv.relationState = reflect.New(h.relationType).Elem()
+	}
+	args := make([]reflect.Value, 2)
+
+	// First check that we have all our required relations
+	incomplete := false
+	for i := 0; i < h.relationType.NumField(); i++ {
+		f := h.relationType.Field(i)
+		data := state.RelationValues[f.Name]
+		// If there's no data, it indicates that the relation is incomplete.
+		if len(data) == 0 {
+			incomplete = true
+		}
+	}
+	if incomplete {
+		return ErrRelationIncomplete
+	}
+
+	// For each field, unmarshal the relation data into the
+	// expected type and call the appropriate setFieldFunc value.
+	restartNeeded := false
+	for i := 0; i < h.relationType.NumField(); i++ {
+		f := h.relationType.Field(i)
```

Comment:
What if inner loop would be turned into a func? It might help to make setRelations more readable.